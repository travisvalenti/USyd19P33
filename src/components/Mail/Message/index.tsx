import React from 'react'
import MessageType from '../../../@types/MessageType'
import Label from '../../../@types/Label'
import AppContext, { AppContextType } from '../../../AppContext'

type Props = {
  message: MessageType,
  isExpanded: boolean,
  onClick: () => void,
  labels: {
    [id: string]: Label
  }
}

type State = {
  attachments: {
    filename: string,
    mimeType: string,
    attachment: string
  }[]
  content?: any
}

class Message extends React.Component<Props, State> {
  static contextType = AppContext
  context!: AppContextType

  constructor (props: Props) {
    super(props)

    this.state = {
      attachments: []
    }
  }

  componentWillUnmount () {
    this.context.timerContext.removeTimer(this.props.message.id)
  }

  async componentDidUpdate (oldProps: Props) {

    // const content = part && atob(part.body.data) // atob decodes a Base64 string
    if (!this.props.isExpanded && oldProps.isExpanded) {
      if (this.context.timerContext.isTimerRunning(this.props.message.id)) {
        this.context.timerContext.removeTimer(this.props.message.id)
      }
    }

    if (this.props.isExpanded && !oldProps.isExpanded) {

      if (this.context.timerContext.isTimerRunning(this.props.message.id)) {
        this.context.timerContext.setTimerToTime(this.props.message.id, 60000)
      } else {
        this.context.timerContext.addTimer(this.props.message.id, () => alert('You\'ve been on that mail for a while'), 60000)
      }

      let part = this.props.message.payload.parts
        .find(part => part.mimeType === 'text/html' || part.mimeType === 'multipart/alternative')
      if (!part) return
      if (part.mimeType === 'multipart/alternative') {
        part = (part as any).parts.find((p: any) => p.mimeType === 'text/html')
      }
      const blob = part && part.body.data
        .split('-').join('+')
        .split('_').join('/')

      let content: string | undefined = ''
      content = blob && atob(blob)

      this.setState({
        content
      })

      // download attchments
      const parts = this.props.message.payload.parts;
      const attachments = this.state.attachments

      await Promise.all(parts.map(async part => {
        if (part.filename && part.filename.length > 0) {
          const attachId: string = part.body.attachmentId

          const gmail = (gapi.client as any).gmail
          return gmail
            .users
            .messages
            .attachments
            .get({
              'id': attachId,
              'messageId': this.props.message.id,
              'userId': 'me'
            })
            .then((response: any) => {
              attachments.push({
                mimeType: part.mimeType,
                filename: part.filename as string,
                attachment: response.result.data
                  .split('-').join('+')
                  .split('_').join('/')
              })
            })
        }
      }))
      this.setState({
        attachments
      })
    }
  }

  render () {
    const from = this.props.message.payload.headers.find((header: any) => header.name === 'From')


    return this.props.message.payload.parts
      ? <div id={this.props.message.id} onClick={() => this.props.onClick()} className={`mailItem ${this.props.message.labelIds.includes('UNREAD') ? 'unread' : ''}`}>
        <div className="mailItemHeader">
          <span>{from && from.value}</span>
          <br />
          <p className="snippet">{this.props.message.snippet}</p>
          {this.props.message.labelIds
            .filter(id => this.props.labels[id].type !== 'system')
            .map(id =>
              <span key={id} className="chip" style={this.props.labels[id].color ? {
                backgroundColor: this.props.labels[id].color!.backgroundColor,
                color: this.props.labels[id].color!.textColor
              } : undefined}>{this.props.labels[id].name}</span>
            )}
        </div>
        {this.props.isExpanded && (<div className="mailItemContent">
          { this.state.attachments.map(attachment => {
            return <div key={attachment.filename} style={{textAlign: 'center'}}>
              <a key={attachment.filename} download={attachment.filename} href={"data:" + attachment.mimeType + ';base64,' + attachment.attachment} onClick={e => e.stopPropagation()}>
                {attachment.filename}
              </a>
              &nbsp;
            </div>
          }) }
          <div className="iframeContainer">
            <iframe title={this.props.message.id} srcDoc={this.state.content} frameBorder="0" seamless onLoad={(frame: any) => frame.target.style.height = (frame.target.contentWindow.document.body.scrollHeight + 25) + 'px'}></iframe>
          </div>
        </div>)}
      </div>
      : null
  }
}

export default Message
