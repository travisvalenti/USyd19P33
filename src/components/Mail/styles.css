.Mail {
  width: 100%;
  height: 100%;
  overflow: auto;
  padding: 1em;
  box-sizing: border-box;
}


.Mail .authButton {
  float: right;
}

.Mail .mailGroup {
  box-shadow:
    2px 2px 3px #ccc,
    -2px 2px 3px #ccc;
  border-radius: 2px;
  margin: 1em auto;
  background : white;
}

.Mail .categoryLabel {
  text-align: center;
  width: 100%;
}

.Mail .mailItem {
  border-bottom: 1px solid #ccc;
  border-radius: 3px;
}

.Mail .mailItem iframe {
  width: 100%;
  margin-top: 5px;
  height: 60vh;
}

.Mail .mailItemHeader {
  padding: 10px;
  background: #eee;
}

.Mail .mailItemContent {
  background: white;
}

.Mail .mailItem.unread .mailItemHeader {
  background: white;
}

.Mail .snippet {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  color: grey;
  font-style: italic;
}

.Mail .chip {
  font-size: 0.6rem;
  font-weight: bold;
  color: white;
  background-color: dodgerblue;
  border-radius: 1em;
  height: 1em;
  margin-right: 0.5em;
  padding: 2px 5px;
  box-sizing: border-box;
}