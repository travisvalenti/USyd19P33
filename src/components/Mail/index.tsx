import React from 'react'

import MessageType from '../../@types/MessageType'
import Label from '../../@types/Label'

import './styles.css'

import LoadingBar from '../ui/LoadingBar'
import Button from '../ui/Button'
import Message from './Message'

type Props = {
  isSignedIn: boolean
}

type State = {
  messages: {
    [key: string]: MessageType,
  },
  isLoading: boolean,
  expandedMessageId?: string,
  errorMessage?: string,
  showRead: boolean,
  importantOnly: boolean
  promotionsExpanded: boolean,
  labels?: {
    [id: string]: Label
  }
}

class Mail extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      messages: {},
      isLoading: false,
      showRead: false,
      importantOnly: false,
      promotionsExpanded: false
    }
  }

  componentDidMount () {
    this.loadLabels()
    this.loadMessages()
  }

  loadLabels = () => {
    const client = gapi.client as any
    client
      .gmail
      .users
      .labels
      .list({
        userId: 'me'
      })
      .then((response: {body: string}) => {
        const labels = Object.fromEntries(
          JSON.parse(response.body).labels
          .map((label: Label) => ([label.id, label]))
        )

        this.setState({ labels })
      })
  }

  loadMessages = () => {
    this.setState({
      isLoading: true
    })
    const client = gapi.client as any
    client
      .gmail
      .users
      .messages
      .list({
        userId: 'me'
      })
      .then((response: { body: string }) => {
        const messages = JSON.parse(response.body).messages
          .reduce((acc: any, cur: any) => {
            return { ...acc, [cur.id]: cur }
          }, {})
        const gapiBatch = gapi.client.newBatch()
        Object.values(messages).forEach((message: any) => {
          const request = (gapi.client as any).gmail.users.messages.get({
            'userId': 'me',
            'id': message.id
          })
          gapiBatch.add(request)
        })
        gapiBatch.then((batchResponse: any) => {
          Object.values(batchResponse.result).forEach((requestResponse: any) => {
            const mail = JSON.parse(requestResponse.body)
            messages[mail.id] = mail
          })
          this.setState({ messages, isLoading: false })
        })

      })
      .catch((error: Error) => {
        console.error(error)
        this.setState({
          isLoading: false,
          errorMessage: error.message || JSON.parse((error as any).body).error.message
        })
      })
  }

  setExpanded = (messageId: string) => {
    this.setState({
      expandedMessageId: messageId === this.state.expandedMessageId ? undefined : messageId
    })
    setTimeout(() => {
      const element = document.getElementById(messageId)
      if (!element) return
      element.scrollIntoView()
    }, 10)
  }

  oneUpdateMessage = (updatedMessage: MessageType) => {
    const messages = { ...this.state.messages }
    if (!messages || !messages[updatedMessage.id]) return

    const request = (gapi.client as any).gmail.users.messages.get({
      'userId': 'me',
      'id': updatedMessage.id
    })

    request.execute((fetchedMessage: MessageType) => {
      messages[updatedMessage.id] = fetchedMessage
      console.log(messages)
      console.log(updatedMessage)
      this.setState({ messages })
    })
  }

  // This function is called whenever the props change or this.setState is called
  render() {
    const promotions = Object.values(this.state.messages)
      .filter(message => message.labelIds.includes('CATEGORY_PROMOTIONS'))
    return (<div className="Mail">
      {this.props.isSignedIn
        ? <Button className="authButton" onClick={() => gapi.auth2.getAuthInstance().signOut()} disabled={this.state.isLoading}>Sign Out</Button>
        : <Button className="authButton" onClick={() => gapi.auth2.getAuthInstance().signIn()} disabled={this.state.isLoading}>Sign In</Button>
      }
      <Button onClick={this.loadMessages} disabled={this.state.isLoading || !this.props.isSignedIn}>Load Messages</Button>
      {this.state.errorMessage && <p style={{ color: 'red', textAlign: 'center' }}>{this.state.errorMessage}</p>}
      { Object.keys(this.state.messages).length > 0 && <>
        <div><input type="checkbox" checked={this.state.showRead} onChange={(event) => this.setState({ showRead: event.target.checked })} /> Show Read</div>
        <div><input type="checkbox" checked={this.state.importantOnly} onChange={(event) => this.setState({ importantOnly: event.target.checked })} /> Only Important</div>
      </> }
      { (this.state.isLoading) && <LoadingBar /> }
      { !!gapi && !!this.state.messages && !!this.state.labels && <>
        {this.props.isSignedIn && (<>
          <div className="mailGroup">
            {Object.values(this.state.messages)
              .filter(message => !message.labelIds.includes('CATEGORY_PROMOTIONS'))
              .map(message =>
              (message.labelIds.includes('UNREAD') || this.state.showRead) &&
              (!this.state.importantOnly || message.labelIds.includes('IMPORTANT')) &&
              <Message key={message.id} message={message} updateMessage={this.oneUpdateMessage} labels={this.state.labels!} isExpanded={this.state.expandedMessageId === message.id} onClick={() => this.setExpanded(message.id)}/>
            )}
          </div>
          {promotions
            .length > 0 && 
            <div className="mailGroup">
            <div className="categoryLabel">
              <h3>Promotions ({promotions.length})</h3>
              <span style={{ color: '#57a5af'}} onClick={() => this.setState({ promotionsExpanded: !this.state.promotionsExpanded})}>
                { this.state.promotionsExpanded ? 'Minimise Promotions' : 'Expand promotions' }
              </span>
            </div>
            { this.state.promotionsExpanded && promotions
                .map(message =>
                  (message.labelIds.includes('UNREAD') || this.state.showRead) &&
                  (!this.state.importantOnly || message.labelIds.includes('IMPORTANT')) &&
                  <Message key={message.id} updateMessage={this.oneUpdateMessage} message={message} labels={this.state.labels!} isExpanded={this.state.expandedMessageId === message.id} onClick={() => this.setExpanded(message.id)} />
                )}
            </div>
          }
          {/* {Object.values(this.state.messages).length !== 0 && <p style={{ textAlign: 'center' }}>There are more messages, but we haven't made a way to load them yet, sorry.</p>} */}
        </>)}

      </>}
    </div>)
  }
}

export default Mail
