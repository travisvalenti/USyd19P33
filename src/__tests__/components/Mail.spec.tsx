import React from 'react'
import { render, cleanup, fireEvent, wait } from '@testing-library/react'

import Mail from '../../components/Mail'

describe('Mail component', () => {
  let gapi: any

  afterEach(cleanup)

  beforeEach(() => {
    (window as any).gapi = {
      load: jest.fn((type: string, callback: () => void) => {
        if (type === 'client:auth2') callback()
      }),
      client: {
        init: jest.fn((options: any) => Promise.resolve()),
        gmail: { users: { messages: {
          list: jest.fn((options: {userId: string}) => Promise.resolve({
            body: `{ "messages": [] }`
          })),
          get: (options: {
            userId: string,
            id:string
          }) => ({})
        }}},
        newBatch: () => ({
          add: (request: {}) => null,
          then: (callback: (response: any) => void) => {
            callback({ result: [] })
          }
        }),

      },
      auth2: {
        getAuthInstance: () => ({
          isSignedIn: {
            get: jest.fn(() => true),
            listen: jest.fn((callback: (isSignedIn: boolean) => void) => {
              setTimeout(() => callback(true), 20)
            })
          }
        })
      }
    }
    gapi = (window as any).gapi
  })

  test('should match snapshot', () => {
    const component = render(<Mail />)
    expect(component.container).toMatchSnapshot()
  })

})