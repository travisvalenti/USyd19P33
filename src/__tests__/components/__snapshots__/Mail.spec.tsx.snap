// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mail component should match snapshot 1`] = `
<div>
  <div
    class="Mail"
  >
    <button
      class="Button authButton"
      disabled=""
    >
      Sign In
    </button>
    <button
      class="Button"
      disabled=""
    >
      Load Messages
    </button>
    <div
      class="LoadingBar"
    >
      <div />
    </div>
  </div>
</div>
`;
